protocol/x/delaymsg/keeper/dispatch.go

```diff
@@ -33,13 +40,21 @@ func DispatchMessagesForBlock(k types.DelayMsgKeeper, ctx sdk.Context) {
 
 		if err = abci.RunCached(ctx, func(ctx sdk.Context) error {
 			handler := k.Router().Handler(msg)
-			_, err := handler(ctx, msg)
-			return err
+			res, err := handler(ctx, msg)
+			if err != nil {
+				return err
+			}
+			// Append events emitted in cached context to `events`.
+			events = append(events, res.GetEvents()...)
+			return nil
 		}); err != nil {
 			k.Logger(ctx).Error("failed to execute delayed message with id %v: %v", id, err)
 		}
 	}
 
+	// Propagate events emitted in cached contexts to current context.
+	ctx.EventManager().EmitEvents(events)
```

Comment:
We should wrap this event emitting logic in `RunCached` util. Does the axelar example [util](https://github.com/axelarnetwork/axelar-core/blob/e9287fd765f7e501238ab85395ba874d5d89b52c/utils/abci.go#L19) work?

Reply:
There were a lot of issues due to sdk message handler creating new EventManager.  The emitted events could be picked up from the result like in: https://github.com/cosmos/cosmos-sdk/issues/13389
or just  propagate events like axelar did and @teddyding is suggesting.I believe Axelar  as well had issues with losing events with their initial `RunCached` implementation: https://github.com/axelarnetwork/axelar-core/pull/1439.

Reply:
When I emit events in `RunCached` like [axelar example](https://github.com/axelarnetwork/axelar-core/blob/e9287fd765f7e501238ab85395ba874d5d89b52c/utils/abci.go#L19), they don't show up any more. The reason is that message handlers create a new event manager and thus the events aren't present in `ctx.EventManager()` in [this line](https://github.com/axelarnetwork/axelar-core/blob/e9287fd765f7e501238ab85395ba874d5d89b52c/utils/abci.go#L36) (The cosmos issue that @dalmirel linked also mentioned the same issue)

I think for `x/delaymsg`'s use case, we have to emit events that are part of handler response